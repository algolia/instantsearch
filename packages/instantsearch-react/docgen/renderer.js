import marked from 'marked';
import {runMode} from 'codemirror/addon/runmode/runmode.node';
import 'codemirror/mode/jsx/jsx';
import escape from 'escape-html';

const highlight = source => {
  let output = '';
  // Even though highlight accepts a lang parameter, force to JSX for now.
  runMode(source, 'jsx', (text, style) => {
    text = escape(text);
    if (!style) {
      output += text;
      return;
    }
    const className = style.split(' ').map(s => `cm-${s}`).join(' ');
    output += `<span class="${className}">${text}</span>`;
  });
  return output;
};

const blockTheme = 'cm-s-default';
const spanTheme = 'cm-s-default';
const renderer = new marked.Renderer();
renderer.code = (code, lang) => {
  return `<pre class="CodeMirror ${blockTheme}"><code>${highlight(code, lang)}</code></pre>`;
};
renderer.codespan = code => {
  return `<code class="CodeMirror ${spanTheme}">${code}</code>`;
};

export default renderer;
