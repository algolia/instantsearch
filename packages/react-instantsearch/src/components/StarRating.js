import React, {PropTypes, Component} from 'react';
import translatable from '../core/translatable';
import classNames from './classNames.js';

const cx = classNames('StarRating');

class StarRating extends Component {
  static propTypes = {
    translate: PropTypes.func.isRequired,
    refine: PropTypes.func.isRequired,
    createURL: PropTypes.func.isRequired,
    min: PropTypes.number.isRequired,
    max: PropTypes.number.isRequired,
    currentRefinement: PropTypes.shape({
      min: PropTypes.number,
      max: PropTypes.number,
    }).isRequired,
    count: PropTypes.arrayOf(PropTypes.shape({
      value: PropTypes.string,
      count: PropTypes.number,
    })).isRequired,
  };

  onClick(min, max, e) {
    e.preventDefault();
    e.stopPropagation();
    if (min === this.props.currentRefinement.min && max === this.props.currentRefinement.max) {
      this.props.refine({min: this.props.min, max: this.props.max});
    } else {
      this.props.refine({min, max});
    }
  }

  buildItem({max, lowerBound, count, translate, createURL, isLowest}) {
    const selected = lowerBound === this.props.currentRefinement.min &&
      max === this.props.currentRefinement.max;
    const disabled = !count;

    const icons = [];
    for (let icon = 0; icon < max; icon++) {
      const iconTheme = icon >= lowerBound ? 'ratingIconEmpty' : 'ratingIcon';
      icons.push(
        <span
          key={icon}
          {...cx(
          iconTheme,
          selected && `${iconTheme}Selected`,
          disabled && `${iconTheme}Disabled`,
        )}
        />);
    }

    // The last item of the list (the default item), should not
    // be clickable if it is selected.
    const isLastAndSelect = isLowest && selected;
    const StarsWrapper = isLastAndSelect ? 'div' : 'a';
    const onClickHandler = isLastAndSelect ? {} : {
      href: createURL({min: lowerBound, max}),
      onClick: this.onClick.bind(this, lowerBound, max),
    };

    return (
      <StarsWrapper {...cx(
        'ratingLink',
        selected && 'ratingLinkSelected',
        disabled && 'ratingLinkDisabled')}
                    disabled={disabled}
                    key={lowerBound}
                    {...onClickHandler}
      >
        {icons}
        <span {...cx(
          'ratingLabel',
          selected && 'ratingLabelSelected',
          disabled && 'ratingLabelDisabled')}>
          {translate('ratingLabel')}
          </span>
        <span> </span>
        <span {...cx(
          'ratingCount',
          selected && 'ratingCountSelected',
          disabled && 'ratingCountDisabled')}>
          {count}
        </span>
      </StarsWrapper>
    );
  }

  render() {
    const {translate, refine, min, max, count, createURL} = this.props;
    const items = [];
    for (let i = max; i >= min; i--) {
      const itemCount = count.reduce((acc, item) => {
        if (item.value >= i) acc = acc + item.count;
        return acc;
      }, 0);
      items.push(this.buildItem({
        lowerBound: i,
        max,
        refine,
        count: itemCount,
        translate,
        createURL,
        isLowest: i === min,
      }));
    }
    return (
      <div {...cx('root')}>
        {items}
      </div>
    );
  }
}

export default translatable({
  ratingLabel: ' & Up',
})(StarRating);
