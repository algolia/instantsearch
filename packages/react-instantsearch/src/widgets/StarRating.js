import connectRange from '../connectors/connectRange.js';
import StarRatingComponent from '../components/StarRating.js';

/**
 * StarRating lets the user refine search results by clicking on stars. The stars are based on the selected attributeName. The underlying rating attribute needs to have from min to max stars.
 * @name StarRating
 * @kind widget
 * @propType {string} attributeName - the name of the attribute in the record
 * @propType {number} min - Minimum value for the rating. When this isn't set, the minimum value will be automatically computed by Algolia using the data in the index.
 * @propType {number} max - Maximum value for the rating. When this isn't set, the maximum value will be automatically computed by Algolia using the data in the index.
 * @propType {{min: number, max: number}} defaultRefinement - Default state of the widget containing the lower bound (end) and the max for the rating.
 * @themeKey ais-StarRating__root - The root component of the widget
 * @themeKey ais-StarRating__ratingLink - The item link
 * @themeKey ais-StarRating__ratingLinkSelected - The selected link item
 * @themeKey ais-StarRating__ratingLinkDisabled - The disabled link item
 * @themeKey ais-StarRating__ratingIcon - The rating icon
 * @themeKey ais-StarRating__ratingIconSelected - The selected rating icon
 * @themeKey ais-StarRating__ratingIconDisabled - The disabled rating icon
 * @themeKey ais-StarRating__ratingIconEmpty - The rating empty icon
 * @themeKey ais-StarRating__ratingIconEmptySelected - The selected rating empty icon
 * @themeKey ais-StarRating__ratingIconEmptyDisabled - The disabled rating empty icon
 * @themeKey ais-StarRating__ratingLabel - The link label
 * @themeKey ais-StarRating__ratingLabelSelected - The selected link label
 * @themeKey ais-StarRating__ratingLabelDisabled - The disabled link label
 * @themeKey ais-StarRating__ratingCount - The link count
 * @themeKey ais-StarRating__ratingCountSelected - The selected link count
 * @themeKey ais-StarRating__ratingCountDisabled - The disabled link count
 * @themeKey ais-StarRating__noRefinement - present when there is no refinement
 * @translationKey ratingLabel - Label value for the rating link
 * @example
 * import React from 'react';
 *
 * import {StarRating, InstantSearch} from '../packages/react-instantsearch/dom';
 *
 * export default function App() {
 *   return (
 *     <InstantSearch
 *       className="container-fluid"
 *       appId="latency"
 *       apiKey="6be0576ff61c053d5f9a3225e2a90f76"
 *       indexName="ikea"
 *     >
 *       <StarRating attributeName="rating" />
 *     </InstantSearch>
 *   );
 * }
 */
export default connectRange(StarRatingComponent);
