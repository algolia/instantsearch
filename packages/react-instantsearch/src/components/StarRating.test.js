/* eslint-env jest, jasmine */

import React from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';

import StarRating from './StarRating';

describe('StarRating', () => {
  it('supports passing max/min values', () => {
    const tree = renderer.create(
      <StarRating
        createURL={() => '#'}
        refine={() => null}
        min={1}
        max={5}
        currentRefinement={{min: 1, max: 5}}
        count={[{value: '1', count: 1},
          {value: '2', count: 2},
          {value: '3', count: 3},
          {value: '4', count: 4},
          {value: '5', count: 5}]}
        canRefine={true}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('applies translations', () => {
    const tree = renderer.create(
      <StarRating
        createURL={() => '#'}
        refine={() => null}
        translations={{
          ratingLabel: ' & Up',
        }}
        min={1}
        max={5}
        currentRefinement={{min: 1, max: 5}}
        count={[{value: '1', count: 1},
          {value: '2', count: 2},
          {value: '3', count: 3},
          {value: '4', count: 4},
          {value: '5', count: 5}]}
        canRefine={true}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  const refine = jest.fn();
  const createURL = jest.fn();
  const starRating = <StarRating
    createURL={createURL}
    refine={refine}
    min={1}
    max={5}
    currentRefinement={{min: 1, max: 5}}
    count={[{value: '1', count: 1},
      {value: '2', count: 2},
      {value: '3', count: 3},
      {value: '4', count: 3},
      {value: '5', count: 0}]}
    canRefine={true}
  />;

  beforeEach(() => {
    refine.mockClear();
    createURL.mockClear();
  });

  it('should create an URL for each row except for the largest: the default selected one', () => {
    const wrapper = mount(starRating);

    expect(createURL.mock.calls.length).toBe(4);
    expect(createURL.mock.calls[0][0]).toEqual({min: 5, max: 5});
    expect(createURL.mock.calls[1][0]).toEqual({min: 4, max: 5});
    expect(createURL.mock.calls[2][0]).toEqual({min: 3, max: 5});
    expect(createURL.mock.calls[3][0]).toEqual({min: 2, max: 5});

    wrapper.unmount();
  });

  it('refines its value on change', () => {
    const wrapper = mount(starRating);
    const links = wrapper.find('.ais-StarRating__ratingLink');
    expect(links.length).toBe(5);

    let selectedLink = wrapper.find('.ais-StarRating__ratingLinkSelected');
    expect(selectedLink.length).toBe(1);

    links.first().simulate('click');

    expect(refine.mock.calls.length).toBe(1);
    expect(refine.mock.calls[0][0]).toEqual({min: 5, max: 5});

    selectedLink = wrapper
      .find('.ais-StarRating__ratingLinkSelected');
    expect(selectedLink).toBeDefined();

    refine.mockClear();

    const disabledLink = wrapper
      .find('.ais-StarRating__ratingLinkDisabled')
      .find('.ais-StarRating__ratingIcon');

    expect(disabledLink.length).toBe(5);
    wrapper.unmount();
  });

  it('should display the right number of stars', () => {
    const wrapper = mount(starRating);
    wrapper
      .find('.ais-StarRating__ratingLink')
      .last()
      .simulate('click');

    const selectedLink = wrapper
      .find('.ais-StarRating__ratingLinkSelected');

    const fullIcon = selectedLink
      .find('.ais-StarRating__ratingIcon');
    const emptyIcon = selectedLink
      .first().find('.ais-StarRating__ratingIconEmpty');

    expect(fullIcon.length).toBe(1);
    expect(emptyIcon.length).toBe(4);
    wrapper.unmount();
  });

  it('clicking on the selected refinement should select the largest range', () => {
    const wrapper = mount(starRating);
    wrapper.setProps({currentRefinement: {min: 5, max: 5}});

    const links = wrapper.find('.ais-StarRating__ratingLink');
    links.first().simulate('click');

    expect(refine.mock.calls.length).toBe(1);
    expect(refine.mock.calls[0][0]).toEqual({min: 1, max: 5});
    wrapper.unmount();
  });
  describe('Panel compatibility', () => {
    it('Should indicate when no more refinement', () => {
      const canRefine = jest.fn();
      const wrapper = mount(
        starRating,
        {
          context: {canRefine},
          childContextTypes: {canRefine: React.PropTypes.func},
        },
      );

      expect(canRefine.mock.calls.length).toBe(1);
      expect(canRefine.mock.calls[0][0]).toEqual(true);
      expect(wrapper.find('.ais-StarRating__noRefinement').length).toBe(0);

      wrapper.setProps({canRefine: false});

      expect(canRefine.mock.calls.length).toBe(2);
      expect(canRefine.mock.calls[1][0]).toEqual(false);
      expect(wrapper.find('.ais-StarRating__noRefinement').length).toBe(1);
    });
  });
});
