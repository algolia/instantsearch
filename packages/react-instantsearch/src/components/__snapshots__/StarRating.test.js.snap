// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarRating applies translations 1`] = `
<div
  className="ais-StarRating__root"
>
  <a
    className="ais-StarRating__ratingLink"
    disabled={false}
    href="#"
    onClick={[Function]}
  >
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingLabel"
    >
       & Up
    </span>
    <span>
       
    </span>
    <span
      className="ais-StarRating__ratingCount"
    >
      5
    </span>
  </a>
  <a
    className="ais-StarRating__ratingLink"
    disabled={false}
    href="#"
    onClick={[Function]}
  >
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIconEmpty"
    />
    <span
      className="ais-StarRating__ratingLabel"
    >
       & Up
    </span>
    <span>
       
    </span>
    <span
      className="ais-StarRating__ratingCount"
    >
      9
    </span>
  </a>
  <a
    className="ais-StarRating__ratingLink"
    disabled={false}
    href="#"
    onClick={[Function]}
  >
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIconEmpty"
    />
    <span
      className="ais-StarRating__ratingIconEmpty"
    />
    <span
      className="ais-StarRating__ratingLabel"
    >
       & Up
    </span>
    <span>
       
    </span>
    <span
      className="ais-StarRating__ratingCount"
    >
      12
    </span>
  </a>
  <a
    className="ais-StarRating__ratingLink"
    disabled={false}
    href="#"
    onClick={[Function]}
  >
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIconEmpty"
    />
    <span
      className="ais-StarRating__ratingIconEmpty"
    />
    <span
      className="ais-StarRating__ratingIconEmpty"
    />
    <span
      className="ais-StarRating__ratingLabel"
    >
       & Up
    </span>
    <span>
       
    </span>
    <span
      className="ais-StarRating__ratingCount"
    >
      14
    </span>
  </a>
  <div
    className="ais-StarRating__ratingLink ais-StarRating__ratingLinkSelected"
    disabled={false}
  >
    <span
      className="ais-StarRating__ratingIcon ais-StarRating__ratingIconSelected"
    />
    <span
      className="ais-StarRating__ratingIconEmpty ais-StarRating__ratingIconEmptySelected"
    />
    <span
      className="ais-StarRating__ratingIconEmpty ais-StarRating__ratingIconEmptySelected"
    />
    <span
      className="ais-StarRating__ratingIconEmpty ais-StarRating__ratingIconEmptySelected"
    />
    <span
      className="ais-StarRating__ratingIconEmpty ais-StarRating__ratingIconEmptySelected"
    />
    <span
      className="ais-StarRating__ratingLabel ais-StarRating__ratingLabelSelected"
    >
       & Up
    </span>
    <span>
       
    </span>
    <span
      className="ais-StarRating__ratingCount ais-StarRating__ratingCountSelected"
    >
      15
    </span>
  </div>
</div>
`;

exports[`StarRating supports passing max/min values 1`] = `
<div
  className="ais-StarRating__root"
>
  <a
    className="ais-StarRating__ratingLink"
    disabled={false}
    href="#"
    onClick={[Function]}
  >
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingLabel"
    >
       & Up
    </span>
    <span>
       
    </span>
    <span
      className="ais-StarRating__ratingCount"
    >
      5
    </span>
  </a>
  <a
    className="ais-StarRating__ratingLink"
    disabled={false}
    href="#"
    onClick={[Function]}
  >
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIconEmpty"
    />
    <span
      className="ais-StarRating__ratingLabel"
    >
       & Up
    </span>
    <span>
       
    </span>
    <span
      className="ais-StarRating__ratingCount"
    >
      9
    </span>
  </a>
  <a
    className="ais-StarRating__ratingLink"
    disabled={false}
    href="#"
    onClick={[Function]}
  >
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIconEmpty"
    />
    <span
      className="ais-StarRating__ratingIconEmpty"
    />
    <span
      className="ais-StarRating__ratingLabel"
    >
       & Up
    </span>
    <span>
       
    </span>
    <span
      className="ais-StarRating__ratingCount"
    >
      12
    </span>
  </a>
  <a
    className="ais-StarRating__ratingLink"
    disabled={false}
    href="#"
    onClick={[Function]}
  >
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIcon"
    />
    <span
      className="ais-StarRating__ratingIconEmpty"
    />
    <span
      className="ais-StarRating__ratingIconEmpty"
    />
    <span
      className="ais-StarRating__ratingIconEmpty"
    />
    <span
      className="ais-StarRating__ratingLabel"
    >
       & Up
    </span>
    <span>
       
    </span>
    <span
      className="ais-StarRating__ratingCount"
    >
      14
    </span>
  </a>
  <div
    className="ais-StarRating__ratingLink ais-StarRating__ratingLinkSelected"
    disabled={false}
  >
    <span
      className="ais-StarRating__ratingIcon ais-StarRating__ratingIconSelected"
    />
    <span
      className="ais-StarRating__ratingIconEmpty ais-StarRating__ratingIconEmptySelected"
    />
    <span
      className="ais-StarRating__ratingIconEmpty ais-StarRating__ratingIconEmptySelected"
    />
    <span
      className="ais-StarRating__ratingIconEmpty ais-StarRating__ratingIconEmptySelected"
    />
    <span
      className="ais-StarRating__ratingIconEmpty ais-StarRating__ratingIconEmptySelected"
    />
    <span
      className="ais-StarRating__ratingLabel ais-StarRating__ratingLabelSelected"
    >
       & Up
    </span>
    <span>
       
    </span>
    <span
      className="ais-StarRating__ratingCount ais-StarRating__ratingCountSelected"
    >
      15
    </span>
  </div>
</div>
`;
